### DISCUSSION TITLE
Inconsistent visibility of protected interface members

### DISCUSSION THREAD

**Discussion by zvrba** [dotnet/csharplang#8269] [General] [answered]
Class implementing interface can access "inherited" `protected static` methods, but not `protected` instance methods.

Given the following:
```
public interface IFactAdapter {
    IReadOnlyList<Term> SchemaFields { get; }
    protected object[] CreateValues2() { ... }
    protected static object[] CreateValues(IReadOnlyList<Term> schemaFields) { ... }
}

class FactAdapter : IFactAdapter
{
    public FactAdapter() {
        var values1 = IFactAdapter.CreateValues(...);  // OK
        var values2 = ((IFactAdapter)this).CreateValues2(); // CS1540 ERROR:
    }
}
```

The full error is CS1540: Cannot access protected member `IFactAdapter.CreateValues2()` via a qualifier of type `IFactAdapter`; the qualifier must be of type `DefaultFactAdapter<T>` (or derived from it).

This is blatantly inconsistent.  I know that a class technically does not "inherit" an interface, but then why is it allowed to access `protected static` interface members? Is it possible to fix this so that also inherited protected members are accessible?

**Comment by HaloFour** [comment]
You've never been able to cast `this` to another type in order to invoke protected members.  I think this is one of those scenarios that fell under the `base(T)` feature, no?  

https://github.com/dotnet/csharplang/issues/2337

```cs
var values2 = base(IFactAdapter).CreateValues2();
```

**Reply by zvrba** [reply]
You're absolutely correct in all of your points.

However, `base(T)` feature has been "in the works" for quite a while now, we don't know when it's going to be ready, and at the same time there's this blatant inconsistency wrt `protected` members on interfaces. The compiler could relax its rules about casts and protected access for just the scenario in question. (While we're waiting for `base(T)`... which is soon 5 and a half years old, so it's like waiting for Godot.)

**Reply by HaloFour** [reply]
I think the problem is that the runtime requires changes in order to support such an invocation, regardless of what syntax would be used.  As soon as `this` is cast to anything you lose the ability to call members that require privileged accessibility.  This is true with inheritance, too.

**Comment by GeeLaw** [comment, marked as answer]
The runtime has always supported this, but to speak from C#, you need some twisted tricks of generics, just like how you **invoke any explicitly implemented interface members.**

```csharp
public interface I
{
  protected void M();
}
public class Base : I
{
  void I.M() { }
}
public class NonSealed : Base
{
  // Both constraints are necessary.
  // : NonSealed ensures NonSealed has access to its inherited protected members.
  // : I makes protected members of I visible, as NonSealed itself has no visible I.M
  //     because the explicit implementation in Base is both private and non-speakable.
  private static void InvokeM<TNonSealed>(TNonSealed that) where TNonSealed : NonSealed, I
  {
    // Of course, this binds to the most-derived reimplementation of I.
    that.M();
  }
  public static void TwiceM(NonSealed obj)
  {
    InvokeM(obj);
    InvokeM(obj);
  }
}
public sealed class Sealed : Base
{
  // The following line fails to compile because Sealed is sealed
  // and C# decided to disallow sealed classes as base class constraints.
  // I don't know whether CLI/IL limits it. I guess it's purely a C# thing.
  // static void InvokeM<TSealed>(TSealed that) where TSealed : Sealed, I
  // {
  //   that.M();
  // }

  public static void TwiceM(Sealed obj)
  {
    // InvokeM(obj);
    // InvokeM(obj);
  }
}
// A detour that works for sealed classes.
public abstract class Helper : Base
{
  private protected abstract void PreventOutsiders();
  // The following method should not be just `protected`.
  // Otherwise, an abstract outsider derived classes
  // would get access to `I.M` on `SealedHotfix`.
  private protected static void InvokeM<THelper>(THelper that) where THelper : Helper, I
  {
    that.M();
  }
}
public sealed class SealedHotfix : Helper
{
  private protected override void PreventOutsiders() { }
  public static void TwiceM(SealedHotfix obj)
  {
    InvokeM(obj);
    InvokeM(obj);
  }
}
```

**Reply by zvrba** [reply]
That's indeed ingenious and twisted. Thanks!

**Reply by martin-steinhoefer** [reply]
Very insightful solution. Thanks for finding out.

But if one follows best practices for explicit interface implementations (needed, because `M` is `protected` and `abstract` in `I`), which states, not to directly implement the member but to delegate the call to an instance member, then the result would look like the following:

```csharp
public interface I
{
    protected void M();
}

public sealed class Impl : I
{
    // explicit implementation of I
    void I.M() => M();

    // instance member implementing the explicit member
    private void M() { }
}
```

With that, there is no need for the trick, just invoke the private member `M`. And there is no need for having a non-sealed class (which may introduce problems on it's own).

**Reply by GeeLaw** [reply]
If the `sealed` class is the first to implement `I` in its hierarchy, or re-implements `I`, then sure, it can provide itself with the convenience. The scenario the detour is necessary is exemplified in the main reply:

- `Base` implements `I`,
- `Derived` derives from `Base` without re-implementing `I` (and wants to be `sealed`),
- `Derived` wants to access `protected` members of `I` on itself (which would take the implementation from `Base`).

Or as shown in the main post, where an implementer class wants to access the default implementation.

Of course, the best practice for `Base` would be to expose alternative entry points to the explicitly implemented members, if the interface is supposed to be otherwise accessible.

```csharp
public interface I { protected void M(); }
public class Base
{
  void I.M() => M();
  protected void M() { }
}
public /* sealed */ class Derived : Base
{
  public void TwiceM() { M(); M(); }
}
```

However, this practice is *not* applicable to the default implementation.

**Reply by GeeLaw** [reply]
Re: the issue with detour for `sealed`. Upon reviewing the solution, there's a leak. The `InvokeM` helper should be made `private protected` instead of `protected` --- because an outsider could abstractly derive from the intermediate class and gets access to `I.M` directly on `SealedHotfix`.

I have retroactively fixed this error.

**Reply by martin-steinhoefer** [reply]
> However, this practice is _not_ applicable to the default implementation.

I missed that case. Thank you for clarification.

**Comment by GeeLaw** [comment]
This reminds me about intersection types and suggests an appealing usage of it. Maybe intersection types could be used as a natural way for C# programmers to express such calls, and also solve this visibility issue in a non-twisted way. Consider the imaginary syntax `T1 & T2` for the compiler to know something is both `T1` and `T2`, i.e., the dynamic type `T` satisfies generic constraint `T : T1, T2, allows ref struct`, and for the compiler to look up identifiers in "a would-be generic method body". Now one can say

```csharp
(this as (FaceAdapter & IFaceAdapter)).CreateValues2();
```

The compiler of course knows `this` is both `FaceAdapter` and `IFaceAdapter`, so the cast compiles to nothing in IL.

In addition, this syntax can be proposed to include value types, to enable expression of `constrained callvirt` on value types in C# without using generics.

Related: #399

**Reply by GeeLaw** [reply]
On a second thought, for value types the syntax is problematic. It’s unclear from the text whether “this” (ref to value type) is being boxed or not. If it’s defined to box, then it’s mostly useless (provides little additional benefit). If it’s defined to not box, the syntax is confusing, because “as” was never used to do a ref (ref to variable) conversion.

**Comment by GeeLaw** [comment]
Additional comment on why the design as-is is the right one.

Suppose `C` implements `I` and `D` derives from `C`. Recall that `D` **need / might not be aware** that `C` actually implements `I`. This happens when `I` is `internal` and `D` lives in a non-friend assembly, or when `I` is nested `private` and `D` is non-nested (or nested elsewhere). Therefore, `C` implementing `I` should not automatically expose any member of `I` to `C`'s subclasses and general users, including the inherited protected members.

This also explains why both constraints are necessary in the solution --- the fact that you can state `I` as a constraint removes block of access to `I`'s members (being in `C` and stating `C` as a constraint further unlocks protected members).
